### ISSUE TITLE
Exception handling issues with "undefined" behavior

### ISSUE THREAD

**Issue by Lyonlancer5** [ppy/osu#750]
# Overview

I'm not completely sure if this qualifies as an "issue" but I tried loading in 0 BPM maps, which I found out about [here.](https://www.youtube.com/watch?v=_sBP7ttRQog)

osu!lazer seems to do different things from what it does with osu!stable (and beta), like to [consume all available memory](https://youtu.be/CjH5_nN_AwA), [throw an OverflowException in a call to `System.TimeSpan.Interval(double, int)`](https://www.youtube.com/watch?v=rh4gHVptLts) or throw an InvocationTargetException.

See the test cases below for a detailed list of information regarding the thrown exceptions.

The reason why I'm not sure because there are no "ranked 0 BPM maps" and that this kind of behavior falls into the "undefined" category.

> `throw new RuntimeException("Sanely handle exceptions?");`

# Test cases

Used map: [senya - Musou no Hana, mapped by -Mo-](https://osu.ppy.sh/s/417106) as seen in the video.
Difficulty "Original (Dear smsh)" is excluded from these test cases as it is a "normal" map

### Without Autoplay:

- 0 Final Full SB: v
- 0 Final Full: v
- 0 Final SB:  v
- 0 Final Slider (No Break): v
- 0 Final Slider: v
- 0 Final Slider 2: v
- 0 Final Fix: **`System.Reflection.TargetInvocationException` was unhandled**
> Message: An unhandled exception of type `System.Reflection.TargetInvocationException` occurred in mscorlib.dll
> Additional information: Exception has been thrown by the target of an invocation.
> Stacktrace: 
![1_stk](https://cloud.githubusercontent.com/assets/7628669/26024224/2811951a-3800-11e7-9682-0d32b6932877.png)
> Case: 
![1](https://cloud.githubusercontent.com/assets/7628669/26024241/768f3878-3800-11e7-8e87-1130f579d644.png)
-----
- 0 Start: v
- 0 Full: v
- 0 First: **`System.OverflowException` was unhandled**
> Message: An unhandled exception of type `System.OverflowException` occurred in mscorlib.dll
> Additional information: TimeSpan overflowed because the duration is too long.
> Stacktrace: 
![0_stk](https://cloud.githubusercontent.com/assets/7628669/26024229/49392974-3800-11e7-8a5a-1d13c4a2153a.png)
> Case: 
![0 start](https://cloud.githubusercontent.com/assets/7628669/26024233/67fdcebe-3800-11e7-8703-e94b470f77a5.png)
-----
- 0 Final Circle: **Completes normally**
- 0 End : Cannot test atm (I don't do standard)
- 0 Single: Cannot test atm (I don't do standard)
- "test: Cannot test atm (I don't do standard)

### With Autoplay (needs more testing):
- 0 Final Circle: **Completes normally**
- All other cases except "test": **Allocates memory until low; now looking for a possible breakpoint in-code to provide more useful info**

### Points taken into consideration:
- Storyboards are not-yet-implemented
- NoFail doesn't work (?)
- The startup message: "This is a development build! Don't expect shit to work perfectly[...]"
- Current hardware setup (3GB RAM is too small)

# Solution (?)
> [T]echnical design decisions... and ...[b]ugs that you have found and are personally willing and able to fix

~~May work on it after setting up the dev env and studying _which does what_ in the code~~
Going to study C# (and .NET) first

**Comment by huoyaoyuan** [comment]
This should be marked as "irregular map", like [Tewi 2B](https://osu.ppy.sh/s/90935) and [Kurikaeshi Hitotsubu](https://osu.ppy.sh/s/56411), and they may use undefined behaviors of the "beatmap system".
Handling irregular map should be very late work for "compatibility".

**Comment by Lyonlancer5** [comment]
> irregular map [...] and they may use undefined behaviors of the "beatmap system"

[Do*](https://osu.ppy.sh/s/227126) [Aspire*](https://osu.ppy.sh/s/484689) [maps*](https://osu.ppy.sh/s/594828) [count*](https://osu.ppy.sh/s/594751) ???
![re1](https://cloud.githubusercontent.com/assets/7628669/26359496/3907c17c-4008-11e7-82d8-cfdb73a0e855.png)
![re2](https://cloud.githubusercontent.com/assets/7628669/26359507/3e5a9b2c-4008-11e7-9325-7d5a3af9a7fe.png)
![re3](https://cloud.githubusercontent.com/assets/7628669/26359683/b2b33c40-4008-11e7-9541-e5d231623d7a.png)

TIL screen size affects the game area and object scaling - relates to autoplay missing out-of-bounds notes when using "standard" resolutions (e.g. `1360x768`) vs "resized" resolutions (e.g. `669x680`)

I can't really back/hunt for anything yet on Bountysource as the ToS states that 18 is the minimum age (plus no experience with _internet commerce_ and _digital wallets_), so...

**Comment by bdach** [comment]
I did some testing, and the situation seems to have gotten better over time (at the time of speaking 2019.1021.0 is the latest release). I cannot reproduce crashes with the mapset in the initial report, all of the Aspire maps I tested seem to be working alright, and I closed one of the related reports above as it no longer reproduced.

Obviously, in the cases of unreasonable values, there are still notifications pertaining to parsing errors showing up when the song is displayed at the carousel (see logs for: [Musou no Hana](https://github.com/ppy/osu/files/3780463/runtime-musou-no-hana.log), [Grenade](https://github.com/ppy/osu/files/3780464/runtime-grenade.log)). Since that is the case, and this seems to be a catch-all issue for all "undefined behaviour" with beatmaps, I think I'm going to leave this open for now - let's just say this comment aims to serve as a status update.
